Document security related options for `ig daemon` (#3322)
inspektor-gadget/inspektor-gadget: docs/reference/verify-assets.mdx

## Verify image-based gadgets

Like our container image, we sign all our image-based gadgets.
The signature are verified by default using Inspektor Gadget public key.

<Tabs groupId="env">
<TabItem value="kubectl-gadget" label="kubectl gadget">
Compared to `ig`, you cannot specify information with regard to verification when calling `kubectl gadget run`:

```bash
$ kubectl gadget run --public-keys="$(cat your-key.pub)" trace_exec
Error: unknown flag: --public-keys
$ kubectl gadget run --verify-image=false trace_exec
Error: unknown flag: --verify-image

> ```

Reviewer: I still think it would suffice to explain that in a paragraph above rather than having another terminal box that invites you to try it out and remember, but okay.

> Instead, all these information are set once at deploy time.
> This avoid them being tampered by any users.
> By default all image-based gadgets are verified using Inspektor Gadget public key:

Reviewer: ```suggestion
When using the Inspektor Gadget on Kubernetes, you have to configure these features at deploy time. This enforces the policies so that users cannot override them.
By default all image-based gadgets are verified using Inspektor Gadget public key:
```

```bash
$ kubectl gadget deploy
...
Inspektor Gadget successfully deployed
$ kubectl gadget run ghcr.io/your_repo/trace_exec -A
Error: fetching gadget information: getting gadget info: rpc error: code = Unknown desc = getting gadget info: initializing and preparing operators: instantiating operator "oci": ensuring image: verifying image "ghcr.io/your_repo/trace_exec": verifying signature: invalid signature when validating ASN.1 encoded signature
$ kubectl gadget run trace_exec -A
K8S.NAMESP… K8S.PODNAME K8S.CONTAI… MNTNS… TIMES… PID    PPID   UID    GID    LOGIN… SESSI… RETVAL ARGS_… UPPER… ARGS_… COMM  ARGS  K8S.…
gadget      gadge…hbh8n gadget      40265… 22867… 53386  53368  0      0      42949… 42949… 0      2      false  35     gadg… /bin… mini…
gadget      gadge…hbh8n gadget      40265… 22867… 53387  53369  0      0      42949… 42949… 0      2      false  35     gadg… /bin… mini…
...
```

You can specify a custom public key at deploy time using `--gadgets-public-keys`:

```bash
$ kubectl gadget deploy --gadgets-public-keys="$(cat your-key.pub)"
...
Inspektor Gadget successfully deployed

> $ kubectl gadget run trace_exec -A

Reviewer: Maybe highlight that this _should_ fail, otherwise it might confuse. Maybe also first do the positive test (running _their_ signed gadget) and then the negative one.

```suggestion
# try to run an official gadget (notice that we didn't include the official Inspektor Gadget public key anymore)
$ kubectl gadget run trace_exec -A
```

Error: fetching gadget information: getting gadget info: rpc error: code = Unknown desc = getting gadget info: initializing and preparing operators: instantiating operator "oci": ensuring image: verifying image "trace_exec": verifying signature: invalid signature when validating ASN.1 encoded signature
$ kubectl-gadget run ghcr.io/your_repo/trace_exec -A
K8S.NAMESP… K8S.PODNAME K8S.CONTAI… MNTNS… TIMES… PID    PPID   UID    GID    LOGIN… SESSI… RETVAL ARGS_… UPPER… ARGS_… COMM  ARGS  K8S.…
gadget      gadge…hbh8n gadget      40265… 22867… 53386  53368  0      0      42949… 42949… 0      2      false  35     gadg… /bin… mini…
gadget      gadge…hbh8n gadget      40265… 22867… 53387  53369  0      0      42949… 42949… 0      2      false  35     gadg… /bin… mini…
...
```

> You can also skip verification at deploy time, note that we do not recommend doing so:

Reviewer: ```suggestion
You can also instruct Inspektor Gadget to skip the verification by disabling it at deploy time. Please note that we do not recommend doing so and that it should only be used for development purposes.
```

## Verify image-based gadgets

Like our container image, we sign all our image-based gadgets.
The signatures are by default verified against the official Inspektor Gadget public key.

<Tabs groupId="env">
<TabItem value="kubectl-gadget" label="kubectl gadget">
Compared to `ig`, you cannot specify information with regard to verification when calling `kubectl gadget run`:

```bash
$ kubectl gadget run --public-keys="$(cat your-key.pub)" trace_exec
Error: unknown flag: --public-keys
$ kubectl gadget run --verify-image=false trace_exec
Error: unknown flag: --verify-image
```

> When using the Inspektor Gadget on Kubernetes, you have to configure these features at deploy time.

Reviewer: ```suggestion
When using Inspektor Gadget on Kubernetes, you have to configure these features at deploy time.
```
Sorry, that one was on me, unfortunately.

WARN[0000] image signature verification is disabled due to using corresponding option
RUNTIME.CONTAINERNAME  PID          UID          GID          MNTNS_ID RET FL… MODE        COMM        FNAME                  TIMESTAMP
```
</TabItem>

<TabItem value="ig-daemon" label="ig daemon">

Like `kubectl-gadget`, you cannot specify information with regard to verification when calling `gadgetctl run`:

```bash
$ gadgetctl run --public-keys="$(cat your-key.pub)" trace_exec --remote-address=$addr
Error: unknown flag: --public-keys
$ gadgetctl run --verify-image=false trace_exec --remote-address=$addr
Error: unknown flag: --verify-image

> ```

Reviewer: Similar to above, I don't think we should give examples that just result in errors like `unknown flag` - those flags  might still be remembered as valid, because it looks like a valid example at first. Explanation alone should suffice, IMHO.

Instead, all these information are set once when starting `ig` as a daemon.

> This avoid them being tampered by any users.

Reviewer: ```suggestion
This avoids them being tampered by any users.
```

By default all image-based gadgets are verified using Inspektor Gadget public key:

```bash
$ IP='your_ip'
$ PORT='your_port'
$ ADDR=tcp://$IP:$PORT

> $ sudo ig daemon --public-keys="$(cat your-key.pub),$(cat inspektor-gadget.pub)" --host=$ADDR

Reviewer: See above: `--host` is not the focus of this example, so IMHO it should be omitted.

...
# Switch to another terminal
$ gadgetctl run ghcr.io/your_repo/trace_exec --remote-address=$ADDR
Error: fetching gadget information: getting gadget info: rpc error: code = Unknown desc = getting gadget info: initializing and preparing operators: instantiating operator "oci": ensuring image: verifying image "ghcr.io/your_repo/trace_exec:latest": the image was not signed by the provided keys: invalid signature when validating ASN.1 encoded signature
$ gadgetctl run trace_exec --remote-address=$ADDR
RUNTIME.CONTAINERNAME    COMM                    PID           TID PCOMM                PPID ARGS         ER… TIMESTAMP
```

> You can specify several public keys:

Reviewer: ```suggestion
You can specify several public keys by separating them with a comma:
```

```bash
$ IP='your_ip'
$ PORT='your_port'
$ ADDR=tcp://$IP:$PORT
$ sudo ig daemon --public-keys="$(cat your-key.pub),$(cat inspektor-gadget.pub)" --host=$ADDR
...
# Switch to another terminal
$ gadgetctl run trace_open:latest --remote-address=$ADDR
RUNTIME.CONTAINER… COMM              PID       TID       UID       GID  FD FNAME                    MODE      ERROR  TIMESTAMP
...
^C
$ gadgetctl run ghcr.io/your-repo/gadget/trace_open:latest --remote-address=$ADDR
RUNTIME.CONTAINERN… TIMESTAMP  PID        UID        GID        MNTNS_ID   ERR        FD         FLAGS      MODE      COMM      FNAME
```

> You can also skip verification at start time, note that we do not recommend doing so:

Reviewer: ```suggestion
You can also make Inspektor Gadget skip the verification on the daemon side. Please note that we do not recommend doing so and that it should only be used for development purposes.
```
